internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -1528,11 +1531,72 @@ func (r *ClusterDeploymentsReconciler) SetupWithManager(mgr ctrl.Manager) error
 		}
 	}
 
+	mapAgentToClusterDeployment := func(a client.Object) []reconcile.Request {
+		log := logutil.FromContext(context.Background(), r.Log).WithFields(
+			logrus.Fields{
+				"agent":           a.GetName(),
+				"agent_namespace": a.GetNamespace(),
+			})
+		agent, ok := a.(*aiv1beta1.Agent)
+		if !ok {
+			log.Errorf("%v was not an Agent", a) // shouldn't be possible
+			return []reconcile.Request{}
+		}
+		if agent.Spec.ClusterDeploymentName != nil {
+			log.Debugf("Map Agent : %s %s CD ref name %s ns %s", agent.Namespace, agent.Name, agent.Spec.ClusterDeploymentName.Name, agent.Spec.ClusterDeploymentName.Namespace)
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+						Name:      agent.Spec.ClusterDeploymentName.Name,
+					},
+				},
+			}
+		} else {
+			return []reconcile.Request{}
+		}
+	}
+
+	agentSpecStatusChangedPredicate := builder.WithPredicates(predicate.Funcs{
+		UpdateFunc: func(updateEvent event.UpdateEvent) bool {
+			oldAgent, ok := updateEvent.ObjectOld.(*aiv1beta1.Agent)
+			if !ok {
+				return false
```

Comment:
Should we log anything if we are having trouble obtaining the agent objects?

Reply:
Nope, because this is not a trouble *obtaining* the object, it would be a trouble with converting the type.

What we do here is we receive `updateEvent.ObjectOld` which we need to map into Agent object. It's purely implementation detail that we need this cast, because even though `updateEvent.ObjectOld` comes from the Watch over `aiv1beta1.Agent`, it doesn't have its type. That's because `updateEvent` itself has `event.UpdateEvent` type. We would need generics or something to have it more (or less) elegant, but as this comes from the framework, I just consume it "as is"